flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/BigQuerySchemaProviderImpl.java

```diff
@@ -0,0 +1,345 @@
+ * the License.
+ */
+
+package com.google.cloud.flink.bigquery.sink.serializer;
+
+import com.google.api.client.util.Preconditions;
+import com.google.api.services.bigquery.model.TableSchema;
+import com.google.cloud.flink.bigquery.common.config.BigQueryConnectOptions;
+import com.google.cloud.flink.bigquery.common.exceptions.BigQueryConnectorException;
+import com.google.cloud.flink.bigquery.common.utils.SchemaTransform;
+import com.google.cloud.flink.bigquery.services.BigQueryServices.QueryDataClient;
+import com.google.cloud.flink.bigquery.services.BigQueryServicesFactory;
+import com.google.cloud.flink.bigquery.services.BigQueryUtils;
+import com.google.protobuf.DescriptorProtos.DescriptorProto;
+import com.google.protobuf.DescriptorProtos.FieldDescriptorProto;
+import com.google.protobuf.DescriptorProtos.FileDescriptorProto;
+import com.google.protobuf.Descriptors.Descriptor;
+import com.google.protobuf.Descriptors.DescriptorValidationException;
+import com.google.protobuf.Descriptors.FileDescriptor;
+import org.apache.avro.LogicalType;
+import org.apache.avro.LogicalTypes;
+import org.apache.avro.Schema;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+
+import javax.annotation.Nullable;
+
+import java.io.Serializable;
+import java.util.EnumMap;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * A class that inherits {@link BigQuerySchemaProvider} deriving {@link Schema} from {@link
+ * TableSchema} sourced from {@link BigQueryConnectOptions}.
+ */
+public class BigQuerySchemaProviderImpl implements Serializable, BigQuerySchemaProvider {
+
+    private final Schema avroSchema;
+    private final DescriptorProto descriptorProto;
+
+    private static final Map<Schema.Type, FieldDescriptorProto.Type> AVRO_TYPES_TO_PROTO;
+    private static final Map<String, FieldDescriptorProto.Type> LOGICAL_AVRO_TYPES_TO_PROTO;
+
+    public BigQuerySchemaProviderImpl(BigQueryConnectOptions connectOptions) {
+        QueryDataClient queryDataClient =
+                BigQueryServicesFactory.instance(connectOptions).queryClient();
+        TableSchema tableSchema =
+                queryDataClient.getTableSchema(
+                        connectOptions.getProjectId(),
+                        connectOptions.getDataset(),
+                        connectOptions.getTable());
+        this.avroSchema = getAvroSchema(tableSchema);
+        this.descriptorProto = getDescriptorSchemaFromAvroSchema(this.avroSchema);
+    }
+
+    public BigQuerySchemaProviderImpl(TableSchema tableSchema) {
+        this.avroSchema = getAvroSchema(tableSchema);
+        this.descriptorProto = getDescriptorSchemaFromAvroSchema(this.avroSchema);
+    }
+
+    public BigQuerySchemaProviderImpl(Schema avroSchema) {
+        this.avroSchema = avroSchema;
+        this.descriptorProto = getDescriptorSchemaFromAvroSchema(this.avroSchema);
+    }
```

Comment:
I had a similar suggestion, but Java expects call to another constructor to be the first statement.
Hence, the third method `BigQuerySchemaProviderImpl(BigQueryConnectOptions connectOptions)` will not be allowed as proposed.

Reply:
ah good point! if you put the construction of the TableSchema object into a static member function, i believe it would work. something like:

```
private static TableSchema getTableSchemaFromOptions(BigQueryConnectOptions connectOptions) {
        QueryDataClient queryDataClient =
                BigQueryServicesFactory.instance(connectOptions).queryClient();
        return queryDataClient.getTableSchema(connectOptions.getProjectId(),
                                                                                           connectOptions.getDataset(),
                                                                                           connectOptions.getTable());
}

public BigQuerySchemaProviderImpl(BigQueryConnectOptions connectOptions) {
        BigQuerySchemaProviderImpl(getTableSchemaFromOptions(connectOptions));
}
```

Reply:
Fixed.

Reply:
thanks! you can resolve this convo